src/psyclone/psyir/tools/dependency_tools.py

```diff
@@ -273,102 +224,443 @@ def _is_loop_suitable_for_parallel(self, loop, only_nested_loops=True):
+        Even `a(n)` and `a(3)` are considered dependent, since no
+        restriction on the variable `n` is made, so `n` could be 3.
+
+        :param index_exp1: the first index expression to compare.
+        :type index_exp1: :py:class:`psyclone.psyir.nodes.Node`
+        :param index_exp2: the second index expression to compare.
+        :type index_exp2: :py:class:`psyclone.psyir.nodes.Node`
+
+        '''
+        sym_maths = SymbolicMaths.get()
+
+        # If the indices can be shown to be never equal, the accesses
+        # to the given subscript are always independent.
+        if sym_maths.never_equal(index_exp1, index_exp2):
+            return True
+
+        # Otherwise we have to conservatively assume that the accesses
+        # are dependent on each other. Additional tests could be added,
+        # e.g. `a(n)` and `a(5)` ... if it should be known that n != 5.
+        return False
+
+    # -------------------------------------------------------------------------
+    @staticmethod
+    def get_dependency_distance(var_name, index_read, index_written):
+        '''Computes the dependency distance between two accesses to the
+        same variable. The distance specifies in how many loop iterations
+        the same memory location would be accessed. E.g. `a(i)=a(i-1)`
+        would have a distance of 1: solving i=(i+di)-1 gives di=1
+        Similarly, `a(i) = a(i) + 1` has a distance of 0, and
+        `a(2*i)=a(2*i+1)` will have a distance of -0.5 (which indicates that
+        with an index that's an integer the accesses are independent). More
+        complex examples can have several solutions, e.g. `a(i*i) = a(i*i)+1`.
+        This will have the distance 0 and `-2*i`. The latter would result
+        in a negative loop index (i+(-2*i)<0), indicating that actually no
+        dependency exists.
+        This function will return the distance, if it is an integer value
+        independent of loop iterations, and None otherwise.
+        Multiple solution are not handled yet, since they are unlikely to
+        be observed in code handled with PSyclone, so they will also
+        return None.
+
+        :param str var_name: name of the one variable used in the two \
+            index expressions.
+        :param index_read: the index expression of the variable read that \
+            is to be compare.
+        :type index_read: :py:class:`psyclone.psyir.nodes.Node`
+        :param index_written: the index expression of the variable written \
+            that is to be compare.
+        :type index_written: :py:class:`psyclone.psyir.nodes.Node`
+
+        :returns: the dependency distance in loop iterations if it is a
+            independent integer value, and None otherwise.
+        :rtype: int or None.
+
+        '''
+        sym_maths = SymbolicMaths.get()
+        sympy_expressions, symbol_map = SymPyWriter.\
+            get_sympy_expressions_and_symbol_map([index_read, index_written])
```

Comment:
A couple of Nemo files fail here with a ParseError. Can we put this inside a try/except and `return None` in case of a SymPy exception (meaning it could not guarantee independence)?